ghcide/bench/lib/Experiments.hs

```diff
@@ -44,93 +42,105 @@ import System.Process
 import System.Time.Extra
 import Text.ParserCombinators.ReadP (readP_to_S)
 
-hygienicEdit :: (?hygienicP :: Position) => TextDocumentContentChangeEvent
-hygienicEdit =
+charEdit :: Position -> TextDocumentContentChangeEvent
+charEdit p =
     TextDocumentContentChangeEvent
-    { _range = Just (Range ?hygienicP ?hygienicP),
-        _rangeLength = Nothing,
-        _text = " "
+    { _range = Just (Range p p),
+      _rangeLength = Nothing,
+      _text = "a"
     }
 
-breakingEdit :: (?identifierP :: Position) => TextDocumentContentChangeEvent
-breakingEdit =
-    TextDocumentContentChangeEvent
-    { _range = Just (Range ?identifierP ?identifierP),
-        _rangeLength = Nothing,
-        _text = "a"
-    }
+data DocumentPositions = DocumentPositions {
+    identifierP :: Maybe Position,
+    stringLiteralP :: !Position,
+    doc :: !TextDocumentIdentifier
+}
 
--- | Experiments have access to these special positions:
--- - hygienicP points to a string in the target file, convenient for hygienic edits
--- - identifierP points to the middle of an identifier, convenient for goto-def, hover and completions
-type HasPositions = (?hygienicP :: Position, ?identifierP :: Position)
+allWithIdentifierPos :: Monad m => (DocumentPositions -> m Bool) -> [DocumentPositions] -> m Bool
+allWithIdentifierPos f docs = allM f (filter (isJust . identifierP) docs)
 
 experiments :: [Bench]
 experiments =
     [ ---------------------------------------------------------------------------------------
-      bench "hover" 10 $ \doc ->
-        isJust <$> getHover doc ?identifierP,
+      bench "hover" 10 $ allWithIdentifierPos $ \DocumentPositions{..} ->
+        isJust <$> getHover doc (fromJust identifierP),
       ---------------------------------------------------------------------------------------
-      bench "edit" 10 $ \doc -> do
-        changeDoc doc [hygienicEdit]
-        waitForProgressDone
+      bench "edit" 10 $ \docs -> do
+        forM_ docs $ \DocumentPositions{..} ->
+          changeDoc doc [charEdit stringLiteralP]
+        waitForProgressDone -- TODO check that this waits for all of them
```

Comment:
This doesn't need to be addressed now, but we should probably change such generic "wait for progress done" to wait for the specific progress message they need, or things can behave unexpectedly when there are different kinds of progress sessions going on simultaneously.

I will probably have to go and do this for my references change to make the tests pass, since the indexing thread for hiedb has its own progress session that runs concurrently with everything else.

Reply:
Yeah, the trouble is finding a suitable message to wait for. And if possible in a backwards compatible way so that the benchmark suite can still be used to compare to older versions